### PR TITLE
test: build config tests

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

## Description

- added unit tests
- wait for #2581

### CODE REVIEW

File: bentoml/_internal/bento/docker/entrypoint.sh [hunk 1]
```diff
@@ -18,6 +18,7 @@ _main() {
 	# Overide the BENTOML_PORT if PORT env var is present. Used for Heroku
 	if [[ -v PORT ]]; then
 		echo "\$PORT is set! Overiding \$BENTOML_PORT with \$PORT ($PORT)"
+		echo "NOTE: Yatai also set \$PORT, hence \$BENTOML_PORT will be overridden."
```
Comment (parano) [inline bentoml/_internal/bento/docker/entrypoint.sh:21]: move this to the inline comments above ? No need to log it

Comment (aarnphm) [reply bentoml/_internal/bento/docker/entrypoint.sh:21]: done

File: bentoml/_internal/bento/docker/entrypoint.sh [hunk 2]
```diff
@@ -11,6 +11,7 @@ _is_sourced() {
 
 _main() {
 	# if first arg looks like a flag
+	echo "${1:0:1}"
```
Comment (parano) [inline bentoml/_internal/bento/docker/entrypoint.sh:14]: Remove?

File: bentoml/_internal/bento/docker/entrypoint.sh [hunk 3]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-set -Eeuo pipefail
+set -Eeuox pipefail
```
Comment (parano) [inline bentoml/_internal/bento/docker/entrypoint.sh:2]: Why adding x here?

Comment (aarnphm) [reply bentoml/_internal/bento/docker/entrypoint.sh]: I was debugging, removed now

File: bentoml/_internal/bento/docker/entrypoint.sh [hunk 4]
```diff
@@ -17,6 +17,7 @@ _main() {
 
 	# Overide the BENTOML_PORT if PORT env var is present. Used for Heroku
 	if [[ -v PORT ]]; then
+		# NOTE: Yatai also set \$PORT, hence \$BENTOML_PORT will be overridden.
```
Comment (parano) [inline bentoml/_internal/bento/docker/entrypoint.sh:20]: Simply update the comment above: 

	# Overide the BENTOML_PORT if PORT env var is present. Used for Heroku **and Yatai**

File: bentoml/_internal/bento/gen.py [hunk 5]
```diff
@@ -21,11 +6,13 @@
 import logging
 from typing import TYPE_CHECKING
 
+import fs
 import attr
 from jinja2 import Environment
 from jinja2.loaders import FileSystemLoader
 
-from ..utils import bentoml_cattr
+from ..utils import bentoml_cattr as cattr
```
Comment (parano) [inline bentoml/_internal/bento/gen.py:14]: let's not do this, it's very confusing - we should make it very clear it is our cattr instance

Comment (parano) [reply bentoml/_internal/bento/gen.py:14]: note that there are still places where we use cattr directly.

File: bentoml/_internal/bento/gen.py [hunk 6]
```diff
@@ -84,28 +83,89 @@ class CustomizableEnv:
...
-def get_template_env(
-    docker_options: DockerOptions, spec: DistroSpec
+def validate_user_template(template: Template, loader: BaseLoader) -> None | t.NoReturn:
+    """
+    Validate all user-defined templates in the given environment.
+    """
+    ctx = template.new_context()
+    if "bento__dockerfile" not in ctx.keys():
+        exc_info = (
+            f"User-defined template `{template}` does not contain `bento__dockerfile`. "
```
Comment (parano) [inline bentoml/_internal/bento/gen.py:113]: why is that a problem?  user should be able to generate everything from scratch without our base templates right?

Comment (parano) [reply bentoml/_internal/bento/gen.py:113]: the user could also be just extending one block? right?

File: requirements/tests-requirements.txt [hunk 7]
```diff
@@ -12,5 +12,5 @@ pytest-asyncio
 pandas
 scikit-learn
 imageio>=2.5.0
-docker
```
Comment (parano) [inline requirements/tests-requirements.txt:15]: why removing docker?

Comment (parano) [reply requirements/tests-requirements.txt:15]: fs is already a bentoml dependency, no need to add it here in tests dependency

File: bentoml/_internal/utils/__init__.py [hunk 8]
```diff
@@ -211,9 +211,20 @@ def copy_file_to_fs_folder(
     dst_folder_path: str = ".",
     dst_filename: t.Optional[str] = None,
 ):
-    """Copy the given file at src_path to dst_fs filesystem, under its dst_folder_path
-    folder with dst_filename as file name. When dst_filename is None, keep the original
+    """
```
Comment (parano) [inline bentoml/_internal/utils/__init__.py:214]: probably no need to change this at all - i'd prefer not using `:code:` for internal docs and Args section doesn't really add any value

File: bentoml/_internal/utils/__init__.py [hunk 9]
```diff
@@ -223,9 +234,22 @@ def copy_file_to_fs_folder(
     fs.copy.copy_file(src_fs, file_name, dst_fs, dst_path)
 
 
-def resolve_user_filepath(filepath: str, ctx: t.Optional[str]) -> str:
-    """Resolve the abspath of a filepath provided by user, which may contain "~" or may
+def resolve_user_filepath(filepath: str, ctx: t.Optional[str]) -> str | t.NoReturn:
+    """
```
Comment (parano) [inline bentoml/_internal/utils/__init__.py:238]: same here

File: bentoml/_internal/bento/gen.py (hunk 6 above)
Comment (aarnphm) [reply bentoml/_internal/bento/gen.py:113]: actually we need they to have our base template to make sure that bento still works

File: requirements/tests-requirements.txt (hunk 7 above)
Comment (aarnphm) [reply requirements/tests-requirements.txt:15]: we don't use docker anymore

File: .gitattributes [hunk 10]
```diff
@@ -1,2 +1,3 @@
 typings/**/*.pyi linguist-generated=true
+tests/unit/_internal/bento/testdata/**/*.yaml linguist-generated=true
```
Comment (parano) [inline .gitattributes]: these YAML files should probably not be linguist-generated=true

File: bentoml/_internal/utils/__init__.py (hunk 8 above)
Comment (aarnphm) [reply bentoml/_internal/utils/__init__.py:214]: it actually helps with lsp hover

File: bentoml/_internal/bento/build_config.py [hunk 11]
```diff
@@ -74,11 +65,13 @@ def _convert_python_version(py_version: t.Optional[str]) -> t.Optional[str]:
     return f"{major}.{minor}"
 
 
-def _convert_cuda_version(cuda_version: t.Optional[str]) -> t.Optional[str]:
+def _convert_cuda_version(
+    cuda_version: t.Optional[t.Union[str, float, int]]
```
Comment (parano) [inline bentoml/_internal/bento/build_config.py:69]: float?
```suggestion
    cuda_version: t.Optional[t.Union[str, int]]
```

File: bentoml/_internal/bento/build_config.py [hunk 12]
```diff
@@ -451,10 +529,10 @@ def write_to_bento(self, bento_fs: "FS", build_ctx: str) -> None:
             # Note: "--generate-hashes" is purposefully not used here because it will
             # break if user includes PyPI package from version control system
             pip_compile_in = bento_fs.getsyspath(
-                fs.path.join(py_folder, "requirements.txt")
+                fs.path.combine(py_folder, "requirements.txt")
```
Comment (parano) [inline bentoml/_internal/bento/build_config.py:532]: Would be nice to document this type of changes in PR in a comment like this:
---

Combine: Join two paths together.

This is faster than [join()](https://docs.pyfilesystem.org/en/latest/reference/path.html#fs.path.join), but only works when the second path is relative, and there are no back references in either path.

https://docs.pyfilesystem.org/en/latest/reference/path.html#fs.path.combine

File: tests/conftest.py [hunk 13]
```diff
@@ -10,16 +14,20 @@
     framework_name="testing", framework_versions={"testing": "v1"}
 )
 
+if TYPE_CHECKING:
+    from _pytest.tmpdir import TempPathFactory
```
Comment (parano) [inline tests/conftest.py:18]: what's the difference between this and `from pytest import TempPathFactory`?

File: bentoml/_internal/bento/build_config.py (hunk 11 above)
Comment (aarnphm) [reply bentoml/_internal/bento/build_config.py:69]: for example `11.6` should works as well

File: tests/conftest.py (hunk 13 above)
Comment (aarnphm) [reply tests/conftest.py:18]: oh the autocompletion imports from _pytest

Comment (aarnphm) [reply tests/conftest.py:18]: Screenshot 2022-06-05 at 17 45 00

**Comment by aarnphm** [comment]
changes for bentoctl is https://github.com/bentoml/BentoML/pull/2547. the same files are kept here for testing purposes. 2547 should get merged first

File: .gitattributes (hunk 10 above)
Comment (aarnphm) [reply .gitattributes]: fixed

**Comment by aarnphm** [comment]
TODO @aarnphm:
- add command args validation for pip-tools

**Comment by aarnphm** [comment]
close this for now, since we have done many manual QA. I will take a look into https://github.com/HypothesisWorks/hypothesis for possible property-based test. Will revisit this after 1.0

cc @ssheng.

**Pull request closed without merging**